### PR TITLE
[java] Fixes for new regression tester errors

### PR DESCRIPTION
**Pull request by adangel**
With the switch to the new projects for regression testing (#3640) new errors in our rules appeared. These errors can be seen here: https://github.com/pmd/pmd/pull/3645#issuecomment-979350144

This PR fixes:

### class cast exception with pattern matching in CloseResource

<details>
  <summary>Stacktrace</summary>

```
Caused by: java.lang.ClassCastException: class net.sourceforge.pmd.lang.java.ast.ASTPatternExpression cannot be cast to class net.sourceforge.pmd.lang.java.ast.ASTLocalVariableDeclaration (net.sourceforge.pmd.lang.java.ast.ASTPatternExpression and net.sourceforge.pmd.lang.java.ast.ASTLocalVariableDeclaration are in unnamed module of loader 'app')
at net.sourceforge.pmd.lang.java.rule.errorprone.CloseResourceRule.getDeclaredTypeOfVariable(CloseResourceRule.java:217)
at net.sourceforge.pmd.lang.java.rule.errorprone.CloseResourceRule.getTypeOfVariable(CloseResourceRule.java:213)
at net.sourceforge.pmd.lang.java.rule.errorprone.CloseResourceRule.lambda$getResourceVariables$0(CloseResourceRule.java:200)
at net.sourceforge.pmd.lang.ast.internal.Filtermap.lambda$filter$1(Filtermap.java:106)
at net.sourceforge.pmd.lang.ast.internal.Filtermap.lambda$thenApply$0(Filtermap.java:63)
at net.sourceforge.pmd.lang.ast.internal.Filtermap.lambda$thenApply$0(Filtermap.java:62)
at net.sourceforge.pmd.lang.ast.internal.TreeWalker.findDescendantsMatching(TreeWalker.java:65)
at net.sourceforge.pmd.lang.ast.internal.TreeWalker.findDescendantsMatching(TreeWalker.java:71)
at net.sourceforge.pmd.lang.ast.internal.TreeWalker.findDescendantsMatching(TreeWalker.java:71)
at net.sourceforge.pmd.lang.ast.internal.TreeWalker.findDescendantsMatching(TreeWalker.java:71)
at net.sourceforge.pmd.lang.ast.internal.TreeWalker.findDescendantsMatching(TreeWalker.java:71)
at net.sourceforge.pmd.lang.ast.internal.TreeWalker.findDescendantsMatching(TreeWalker.java:71)
at net.sourceforge.pmd.lang.ast.internal.TreeWalker.findDescendantsMatching(TreeWalker.java:95)
at net.sourceforge.pmd.lang.ast.internal.AxisStream$FilteredDescendantStream.toListImpl(AxisStream.java:264)
at net.sourceforge.pmd.lang.ast.internal.AxisStream.toList(AxisStream.java:107)
at net.sourceforge.pmd.lang.java.rule.errorprone.CloseResourceRule.getResourceVariables(CloseResourceRule.java:202)
at net.sourceforge.pmd.lang.java.rule.errorprone.CloseResourceRule.checkForResources(CloseResourceRule.java:165)
at net.sourceforge.pmd.lang.java.rule.errorprone.CloseResourceRule.visit(CloseResourceRule.java:159)
```
</details>

Sample file: https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/it/resources-noncompilable/com/google/checkstyle/test/chapter5naming/rule527localvariablenames/InputPatternVariableNameEnhancedInstanceofTestDefault.java

Fix: Do a explicit instanceof check, handle TypePattern

### class cast exception with records in ClassNamingConventions

<details>
  <summary>Stacktrace</summary>

```
Caused by: java.lang.ClassCastException: class net.sourceforge.pmd.lang.java.ast.ASTRecordDeclaration cannot be cast to class net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration (net.sourceforge.pmd.lang.java.ast.ASTRecordDeclaration and net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration are in unnamed module of loader 'app')
at net.sourceforge.pmd.lang.java.rule.internal.JavaRuleUtil.isUtilityClass(JavaRuleUtil.java:254)
at net.sourceforge.pmd.lang.java.rule.codestyle.ClassNamingConventionsRule.kindDisplayName(ClassNamingConventionsRule.java:94)
at net.sourceforge.pmd.lang.java.rule.codestyle.ClassNamingConventionsRule.kindDisplayName(ClassNamingConventionsRule.java:22)
at net.sourceforge.pmd.lang.java.rule.codestyle.AbstractNamingConventionRule.checkMatches(AbstractNamingConventionRule.java:64)
at net.sourceforge.pmd.lang.java.rule.codestyle.ClassNamingConventionsRule.visit(ClassNamingConventionsRule.java:70)
```
</details>

Sample file: https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableCheckRecords.java

Fix: `JavaRuleUtil.isUtilityClass` now only considers "regular classes".

### NullpointerException with record constructor in ImplicitSwitchFallthrough

<details>
  <summary>Stacktrace</summary>

```
Caused by: java.lang.NullPointerException
at java.base/java.util.Objects.requireNonNull(Objects.java:221)
at net.sourceforge.pmd.lang.java.rule.internal.DataflowPass$DataflowResult.switchBranchFallsThrough(DataflowPass.java:241)
at net.sourceforge.pmd.lang.java.rule.errorprone.ImplicitSwitchFallThroughRule.visit(ImplicitSwitchFallThroughRule.java:39)
```
</details>

Sample file: https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityRecords.java

Fix: The Dataflow impl now also visits constructors of records.

### Parser error when parsing permits list of a sealed class with qualified names

<details>
  <summary>Stacktrace</summary>

```
Caused by: java.lang.AssertionError: Invalid simple name 'Square.StaticClass'
at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceType.assertSimpleNameOk(ASTClassOrInterfaceType.java:99)
at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceType.setImage(ASTClassOrInterfaceType.java:89)
at net.sourceforge.pmd.lang.java.ast.JavaParserImpl.TypeName(JavaParserImpl.java:2208)
at net.sourceforge.pmd.lang.java.ast.JavaParserImpl.PermittedSubclasses(JavaParserImpl.java:846)
```
</details>

Sample file: https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/modifierorder/InputModifierOrderSealedAndNonSealed.java

Fix: We reuse the `ClassOrInterfaceType` production instead of using a custom `TypeName` production. This TypeName production was only used for the permits list.

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 51 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 33 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/a0787caf3188430db966c27e0d93b4d9/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 45901 violations,<br />
introduces 25677 new violations, 5 new errors and 0 new configuration errors,<br />
removes 182039 violations, 28 errors and 3 configuration errors.<br />
<a href="https://chunk.io/pmd/a0787caf3188430db966c27e0d93b4d9/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloseResourceRule.java [hunk 1]
```diff
@@ -214,8 +215,15 @@ private TypeNode getTypeOfVariable(ASTVariableDeclaratorId var) {
     }
 
     private TypeNode getDeclaredTypeOfVariable(ASTVariableDeclaratorId var) {
-        ASTLocalVariableDeclaration localVar = (ASTLocalVariableDeclaration) var.getParent().getParent();
-        return localVar.getTypeNode(); // note: can be null, if type is inferred (var)
+        JavaNode exprOrDecl = var.getParent().getParent();
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloseResourceRule.java:218]: Isn't there already `ASTVariableDeclaratorId::getTypeNode` that does the same thing as this entire function ?

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloseResourceRule.java:218]: Found it https://docs.pmd-code.org/apidocs/pmd-java/7.0.0-SNAPSHOT/net/sourceforge/pmd/lang/java/ast/ASTVariableDeclaratorId.html#getTypeNode()

Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloseResourceRule.java]: :+1: it works, thanks for the suggestion!

**Comment by adangel** [comment]
The regression report now looks good. The strange "81 changed violations" are gone now. Not sure what happened there - it was one file in spring on which PMD reported different line numbers (offset of 2).

I'll merge this now.

**Pull request merged**